### PR TITLE
[GraphQL] Add query logging when timing out

### PR DESCRIPTION
**Pull request by stefan-mysten**
## Description 

This PR enables logging of queries that time out. The logging uses the `error` macro together with a new error code, `REQUEST_TIMEOUT`.

Note, that the timeout extension function was switched to `parse_query` + `execute`. Execute one takes care of the time, parse query takes care of stringifying up the query (including resolving any passed, fragments, etc).

## Test Plan 

Existing tests, manually checking for errors when setting a small (500ms) timeout.
```bash
2024-02-28T23:25:23.373001Z ERROR sui_graphql_rpc::extensions::timeout: Query: {
  transactionBlocks(first: 50) {
    nodes {
      digest
    }
  }
} query_id=97a7adca-52dc-4e4e-9cdc-bfdaaf1d65ae session_id=127.0.0.1:52232 error_code="REQUEST_TIMEOUT"
```

```
2024-02-29T02:35:24.740291Z ERROR sui_graphql_rpc::extensions::timeout: query_id=44d9f1ec-e4df-488c-bd2a-a0b31d5bb203 session_id=127.0.0.1:59043 error_code="REQUEST_TIMEOUT" query=query { a:transactionBlocks(first: 10) { nodes { digest } } b:transactionBlocks(first: 50) { nodes { digest } } }
```

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/server/builder.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/server/builder.rs:357]: would this capture variables? we have similar logic in `logger.rs`, where we log `ctx.stringify_execute_doc(&document, variables)` if specified in config. But for the timeout extension i think it'd need to hook into `request` or `execute` so we can correctly time out while waiting for the response. I'm not sure if there's a way to access that same data in the `execute` hook, or if it would be a code smell to have `logger` somehow populate global context with that information for the `timeout` extension to use...

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:49]: Yes, we can do that.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:43]: In general, if these are not in the context the whole thing will panic. However, given that this extension is already at the execute phase, these should exist. Indeed, we could use `ctx.data_opt` just to make sure, but in the logger we use the same pattern so I kept it consistent.

File: crates/sui-graphql-rpc/src/server/builder.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/server/builder.rs:357]: Good point, let me think more and see what I can do.

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/server/builder.rs:357]:
```
sui_graphql_rpc::extensions::timeout: query_id=44d9f1ec-e4df-488c-bd2a-a0b31d5bb203 session_id=127.0.0.1:59043 error_code="REQUEST_TIMEOUT" query=query { a:transactionBlocks(first: 10) { nodes { digest } } b:transactionBlocks(first: 50) { nodes { digest } } }
```

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (emmazzz) [inline crates/sui-graphql-rpc/src/extensions/timeout.rs:56]: IIUC `execute` is called once per top level query and a query can include many fields/subqueries. If a query times out, we won't know where exactly (i.e. while resolving which field) the timeout happens. I'm wondering if it makes sense to add the `timeout` and logging to `resolve` instead. This way we have the path information as well and a more precise target for optimization.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/extensions/timeout.rs:54]: The use of `take` here is a bit strange here -- why are we taking it instead of borrowing it, and why are we doing it here vs once we know that an error has happened?

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/extensions/timeout.rs:12]: I notice we're using tokio's `Mutex` here rather than the `std` one. Tokio has some [docs](https://doc.servo.org/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) on the pro's/con's.

Maybe this is something you've looked at but I wanted to flag because I think in this case the `std` version may be more appropriate?

File: crates/sui-graphql-rpc/src/server/builder.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/server/builder.rs:320]: Please create a constructor function for the timeout -- this is leaking an implementation detail at the moment.

**Review by amnn — COMMENTED** [review]
I like the idea overall! Just had some questions about implementation details.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:56]: Happy to have it either way. My thought was that it might make more sense to know all the query, such that we can tear it down to parts that do not work during our debugging. It might give us a bit more context of what is the need and what is the user trying to achieve, rather than just fixing having a sub-query that times out.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:41]: That is a good question. From my tests, that shouldn't happen, but honestly, I do not know how the precedence of executing these extensions functions. The `query_limits_checker` checks on the query payload size in the same `parse_query`, but I think it only uses the query string and variables, and this extra `query` we use for timeouts is not added to that pile.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:54]: Regarding the 2nd part of the question, putting this into the `unwrap_or_else` closure will not work due to the need for calling `await`.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:41]: I think the order in which we add them to the schema is the order in which they are executed. Idk how clean it would be, but we could have the `query_limits_checker` add the stringified doc for the timeout extension to pick up

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:12]: Thanks, I switched to std Mutex. I only used tokio as we're using in some other parts of the code...

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:54]: Switching to std::sync::Mutex fixed this.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:61]: Yes, I think we can use our own `graphql_error_response` that wraps `Response::from_errors` and includes the `error_code`.

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:61]: Btw, I am not sure I follow why it would be a breaking change? What we return now is the following. We add error_code = "UNKNOWN" to all errors that do not have an `error_code` set, so instead this would become REQUEST_TIMEOUT. I'm probably missing something..
```
2024-02-29T17:48:37.265024Z  INFO sui_graphql_rpc::extensions::logger: [Response] message=Request timed out. Limit: 0.5s query_id=8cddf7c6-370a-4de5-92fe-bee48dc23bf6 session_id=127.0.0.1:61093 error_code="UNKNOWN"
```

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs]: Not sure I follow. This is my mental model:
Extension 1: Timeout
Extension 2: QueryLimitsChecker
Extension 3: Logger

Ext 1 + Ext 2  + Ext 3 ==> all call `parse_query`
Ext 1 + Ext 3 ==> call `execute`. 

I do not understand/know in which order is `parse_query` called. I assume it is called 3 times, once for each extension.

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs]: Not sure how to find this, but my intuition says that all 3 need to finish, so they the next call (resolve) will have to wait until these are done. If one of them errors, then the whole thing errors and it returns.

And finally, the query payload will not be impacted by this, so it should all be good.

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (emmazzz) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:56]: Sounds good!

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (amnn) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs]: Yes, all three parse queries need to finish before `async_graphql` moves on to the next phase, but the question is whether we can guarantee that the the limits check happens before the the timeout check, always, so that we don't try and copy a query that we will shortly find out is too big).

File: crates/sui-graphql-rpc/src/extensions/timeout.rs
Comment (amnn) [reply crates/sui-graphql-rpc/src/extensions/timeout.rs:61]: Changing the error code in the log message is not a breaking change, but changing the error code that we return to users would be.

**Review by amnn — APPROVED** [review]
Changes now look good (pending CI going all green!) But do double check the order of operations on extensions before landing (that query limits happens before timeout).

**Comment by stefan-mysten** [comment]
> Changes now look good (pending CI going all green!) But do double check the order of operations on extensions before landing (that query limits happens before timeout).

Yes, we  (Will and myself) figured out that order of extensions matter. Query limits checker is before Timeout, so we're good here. 

The CI issue was fixed too.

**Pull request merged**
